src/libraries/System.Console/src/System/IO/StdInReader.cs

```diff
@@ -363,5 +375,21 @@ private unsafe ConsoleKeyInfo ReadKey()
 
         /// <summary>Gets whether there's input waiting on stdin.</summary>
         internal static bool StdinReady => Interop.Sys.StdinReady();
+
+        private void Echo(char c)
+        {
+            ReadOnlySpan<char> chars = [ c ];
+            Span<byte> bytes = stackalloc byte[32]; // 32 bytes seems ample
+            int bytesWritten = _echoEncoder.GetBytes(chars, bytes, flush: false);
```

Comment:
ASCII should be very common, we should add an optimized code path for it. Sth like this:

```cs
Span<byte> bytes;
if (Ascii.IsValid(c))
{
    byte b = (byte)c;
    bytes = new Span<byte>(in b);
}
else
{
    bytes = stackalloc byte[32]; // 32 bytes seems ample
    int bytesWritten = _echoEncoder.GetBytes(chars, bytes, flush: false);
    if (bytesWritten == 0)
    {
        return;
    }
    bytes = bytes.Slice(0, bytesWritten);
}

lock (Console.Out) // synchronize with other writers
{
    ConsolePal.Write(StdInHandle, bytes);
}
```

Reply:
> if (Ascii.IsValid(c)) byte b = (byte)c;

I am not an encoding expert. Is this valid for any encoding?

Reply:
> I am not an encoding expert. Is this valid for any encoding?

@GrabYourPitchforks @tarekgh Could you please confirm/deny?

Reply:
> if (Ascii.IsValid(c)) byte b = (byte)c;

That's valid.

`Ascii.IsValid` returns `true` if `c <= 127` only, thus the byte-cast is valid and doesn't truncate anything, etc.

> for any encoding?

Side note: If there's any encoding that differs from ASCII in the 0-7th LSB then .NET would have a bigger problem with lots of broken code.

Reply:
Be careful with that. There are some double-bytes encodings which have leading and trailing bytes. Some encodings have trailing bytes in the ASCII range. Like:

https://github.com/dotnet/runtime/blob/6ffa54863c810d0a0fe588a107cb44415709355f/src/libraries/System.Text.Encoding.CodePages/src/System/Text/GB18030Encoding.cs#L23

Another cases, some encodings use escape sequence in the encoding which can have ASCII characters like:

https://github.com/dotnet/runtime/blob/6ffa54863c810d0a0fe588a107cb44415709355f/src/libraries/System.Text.Encoding.CodePages/src/System/Text/ISO2022Encoding.cs#L326

Reply:
Thanks for these insights!

But `if (Ascii.IsValid(c)) byte b = (byte)c;` is valid when `c` is a `char` (UTF-16)?

Reply:
Yes, if you are doing encoding, then the code is fine. I was just pointing out to be careful if there is any code later is decoding bytes and using similar optimization.